spring-cloud-gcp-bigquery/src/main/java/org/springframework/cloud/gcp/bigquery/integration/BigQueryTemplate.java

```diff
@@ -0,0 +1,196 @@
+import java.time.Duration;
+import java.util.concurrent.ScheduledFuture;
+
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.FormatOptions;
+import com.google.cloud.bigquery.Job;
+import com.google.cloud.bigquery.JobInfo.WriteDisposition;
+import com.google.cloud.bigquery.JobStatus.State;
+import com.google.cloud.bigquery.TableDataWriteChannel;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.WriteChannelConfiguration;
+
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.util.StreamUtils;
+import org.springframework.util.concurrent.ListenableFuture;
+import org.springframework.util.concurrent.SettableListenableFuture;
+
+/**
+ * Helper class which simplifies common operations done in BigQuery.
+ *
+ * @author Daniel Zou
+ */
+public class BigQueryTemplate {
+
+	private final BigQuery bigQuery;
+
+	private final String datasetName;
+
+	private final TaskScheduler taskScheduler;
+
+	private boolean autoDetectSchema = true;
+
+	private WriteDisposition writeDisposition = WriteDisposition.WRITE_APPEND;
+
+	private Duration jobPollInterval = Duration.ofSeconds(2);
+
+	public BigQueryTemplate(BigQuery bigQuery, String datasetName, TaskScheduler taskScheduler) {
+		this.bigQuery = bigQuery;
+		this.datasetName = datasetName;
+		this.taskScheduler = taskScheduler;
+	}
+
+	/**
+	 * Sets whether BigQuery should attempt to autodetect the schema of the data when loading
+	 * data into an empty table for the first time.
+	 * @param autoDetectSchema whether data schema should be autodetected from the structure of
+	 * 		the data. Default is true.
+	 */
+	public void setAutoDetectSchema(boolean autoDetectSchema) {
+		this.autoDetectSchema = autoDetectSchema;
+	}
+
+	/**
+	 * Sets the {@link WriteDisposition} which specifies how data should be inserted into
+	 * BigQuery tables.
+	 * @param writeDisposition whether data should be appended or truncated to the BigQuery table.
+	 * 		Default is {@code WriteDisposition.WRITE_APPEND} to append data to a table.
+	 */
+	public void setWriteDisposition(WriteDisposition writeDisposition) {
+		this.writeDisposition = writeDisposition;
+	}
+
+	/**
+	 * Sets the {@link Duration} amount of time to wait between successive polls on the status of
+	 * a BigQuery job.
+	 * @param jobPollInterval the {@link Duration} poll interval for BigQuery job status polling
+	 */
+	public void setJobPollInterval(Duration jobPollInterval) {
+		this.jobPollInterval = jobPollInterval;
+	}
+
+	/**
+	 * Writes data to a specified BigQuery table.
+	 *
+	 * @param tableName name of the table to write to
+	 * @param dataFile data file to write to BigQuery
+	 * @param dataFormatOptions the format of the data in the data file
+	 * @return {@link ListenableFuture} containing the BigQuery Job indicating completion of operation
+	 * @throws IOException if the data fails to be loaded
+	 */
+	public ListenableFuture<Job> writeDataToTable(
```

Comment:
Do we really need this convenience method and the one below?
It doesn't seem to be used yet.
You can also argue that accepting a Spring `Resource` might be useful, or more useful.

Reply:
Done. Removed the other helpers. I prefer `InputStream` though over Resource since `InputStream` is still more versatile and user just has to call `resource.getInputStream()` to go back from resource to input stream.